src/Tools/Common/DsRouterProcessLauncher.cs

```diff
@@ -0,0 +1,110 @@
+
+namespace Microsoft.Internal.Common.Utils
+{
+       internal sealed partial class DsRouterProcessLauncher
+    {
+        [GeneratedRegex(@"pid=\d+")]
+        private static partial Regex ProcessIdRegex ();
+
+        private Process _childProc;
+        private Task _stdOutTask = Task.CompletedTask;
+        private Task _stdErrTask = Task.CompletedTask;
+        internal static DsRouterProcessLauncher Launcher = new();
+        private bool _processStarted;
+
+        private static async Task<int> ReadAndIgnoreToPidAsync(StreamReader streamToIgnore, CancellationToken cancelToken)
+        {
+            Memory<char> memory = new char[4096];
+            while (await streamToIgnore.ReadAsync(memory, cancelToken).ConfigureAwait(false) != 0)
+            {
+                if (ProcessIdRegex().Match(memory.Span.ToString()) is System.Text.RegularExpressions.Match match && match.Success)
+                {
+                    return int.Parse(match.Value.Substring(4));
+                }
+            }
+            return -1;
+        }
+
+        private static async Task ReadAndIgnoreAllStreamAsync(StreamReader streamToIgnore, CancellationToken cancelToken)
+        {
+            Memory<char> memory = new char[4096];
+            while (await streamToIgnore.ReadAsync(memory, cancelToken).ConfigureAwait(false) != 0)
+            {
+            }
+        }
+
+        private bool HasChildProc => _childProc != null;
+
+        private Process ChildProc => _childProc;
+
+        public int Start(string routername, CancellationToken ct)
+        {
+            string toolsRoot = System.IO.Path.GetDirectoryName(System.Environment.ProcessPath);
+            string dotnetDsrouterTool = "dotnet-dsrouter";
+
+            if (!string.IsNullOrEmpty(toolsRoot))
+            {
+                dotnetDsrouterTool = Path.Combine(toolsRoot, dotnetDsrouterTool);
+            }
+
+            _childProc = new Process();
+
+            _childProc.StartInfo.FileName = dotnetDsrouterTool;
+            _childProc.StartInfo.Arguments = routername;
+            _childProc.StartInfo.UseShellExecute = false;
+            _childProc.StartInfo.UseShellExecute = false;
+            _childProc.StartInfo.RedirectStandardOutput = true;
+            _childProc.StartInfo.RedirectStandardError = true;
+            _childProc.StartInfo.RedirectStandardInput = true;
+            try
+            {
+                _childProc.Start();
+                _processStarted = true;
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"An error occurred trying to start process '{_childProc.StartInfo.FileName}' with working directory '{System.IO.Directory.GetCurrentDirectory()}'");
+                Console.WriteLine($"{e.Message}");
+            }
+
+            _stdErrTask = ReadAndIgnoreAllStreamAsync(_childProc.StandardError, ct);
+            Task<int> pidtask = ReadAndIgnoreToPidAsync(_childProc.StandardOutput, ct);
```

Comment:
Are there other reasons to read up to pid from the process output than just getting the pid? If not you could get the process id through _childProc.Id and on failure in this method pid can be set to -1, since that seems to be the only thing other code checks.

Reply:
no, that's just me bing stupid

Reply:
that was to make sure the router was actually running. I replaced that by a 1sec wait delay, and check that it hasn't exited